### PR TITLE
Added EVENT_RULE_ENGINE flag based opt-in

### PR DESCRIPTION
**Pull request by zaingz**
## Motivation
This PR adds support for a Python-based event rule engine as an alternative to the Java implementation. 
This addresses installation issues some users face with the Java event ruler (GH-11299), particularly 
in environments with proxy setups or SSL certificate issues.

## Changes
- Created new utility `event_matcher.py` to centralize event matching logic
- Supports both string (EventBridge) and dict (ESM/Pipes) inputs
- Maintains Java engine as default but allows configuration via EVENT_RULE_ENGINE
- Updated references in:
  - events/provider.py
  - events/v1/provider.py
  - lambda_/event_source_mapping/pollers/poller.py
- Added unit test

## Testing
1. Run the new test suite:
```bash
pytest tests/unit/utils/test_event_matcher.py

### CODE REVIEW

File: localstack-core/localstack/utils/event_matcher.py
Comment (joe4dev) [inline localstack-core/localstack/utils/event_matcher.py:37]: In the long run, we could consider unifying the usages in ESM and Pipes to use strings. Hence, unifying everything in a single helper makes sense here 👍

File: localstack
Comment (joe4dev) [inline localstack:1]: Adding `localstack` as a submodule here might be accidental?

File: localstack-core/localstack/utils/event_matcher.py
Comment (joe4dev) [inline localstack-core/localstack/utils/event_matcher.py:1]: nit: file path comment seems redundant

File: localstack-core/localstack/utils/event_matcher.py
Comment (joe4dev) [inline localstack-core/localstack/utils/event_matcher.py:6]: Minor: We could consider grouping the Java and Python implementations because they are now spread across three different directories.

**Review by joe4dev — COMMENTED** [review]
Nice centralization and addition of unit tests (which we probably under-utilize) 👍 

Feel free to add the change type label and self-assign the PR.

Can we trigger the CI run here by authenticating with CircleCI?

File: localstack-core/localstack/utils/event_matcher.py
Comment (zaingz) [reply localstack-core/localstack/utils/event_matcher.py:32]: think the current approach (Python as default) might be more appropriate for LocalStack because:
- It reduces dependencies for basic usage
- Makes debugging easier
- Follows the principle of least surprise (Python users expect Python implementation by default)

Comment (joe4dev) [reply localstack-core/localstack/utils/event_matcher.py:38]: I agree with all of these arguments.

The problem is that the current python implementation fails 19/64 test cases and causes many issues and surprises for users (based on support requests and GH issues). From a customer's perspective, we tend to favor AWS parity over performance here. Therefore, we have this open PR intending to switch the defaults LocalStack-wide https://github.com/localstack/localstack/pull/11807
An overview testing table is in our internal backlog item, "Switch default for event ruler to Java."

What do you think?

File: localstack-core/localstack/services/events/provider.py
Comment (joe4dev) [inline localstack-core/localstack/services/events/provider.py:1443]: To avoid regressions in the EventStudio project, I think we should avoid this change.
Ultimately, there should be comments for every instrumentation and/or a list of instrumented places because it's currently impossible to keep track of them.

/cc @maxhoheiser @dominikschubert

File: localstack-core/localstack/utils/event_matcher.py
Comment (joe4dev) [inline localstack-core/localstack/utils/event_matcher.py:41]: We noticed some bigger issues related to the Java-based implementation concerning the LocalStack shutdown behavior and decided it's better to switch the default to Python here. Hence, I would suggest to flip the conditional back to your original version.

I'm sorry for this back and forth 😢

**Review by joe4dev — COMMENTED** [review]
Could you please rebase to resolve the conflicts with master? Thank you 🙏

File: tests/aws/services/events/test_archive_and_replay.py
Comment (joe4dev) [inline tests/aws/services/events/test_archive_and_replay.py:367]: This condition should be on the `EVENT_RULER` config because now the test gets always skipped

File: localstack-core/localstack/services/lambda_/event_source_listeners/utils.py
Comment (joe4dev) [inline localstack-core/localstack/services/lambda_/event_source_listeners/utils.py:1]: Can we move `utils.py` to the `event_source_mapping` directory? I think it's cleaner than re-introducing the legacy directory

**Review by joe4dev — COMMENTED** [review]
LGTM after moving the restored utils file and fixing the test skip conditional 👏 

Let's wait for a full 🟢 test run of localstack-ext (see link in Notion ticket), then we're good to merge 🚀

File: tests/aws/services/events/test_archive_and_replay.py
Comment (joe4dev) [reply tests/aws/services/events/test_archive_and_replay.py:367]: ideally, this should be:

```
    @pytest.mark.skipif(
        condition=config.EVENT_RULE_ENGINE == "python",
        reason="Not supported with Python-based rule engine",
    )
```

File: tests/unit/utils/test_event_matcher.py
Comment (joe4dev) [inline tests/unit/utils/test_event_matcher.py:33]: @dominikschubert @dfangl Should we avoid any JPype-starting monkey-patched tests in the LS CI?
I'm somewhat worried this will blow up once we combine it with other usage (e.g., SFN tests)?

This shouldn't block the merge here. I'm happy to tackle this in a follow-up because we have several other usages.

File: tests/aws/services/events/test_archive_and_replay.py
Comment (joe4dev) [reply tests/aws/services/events/test_archive_and_replay.py:367]: (the reason currently contradicts the condition ;)

**Review by joe4dev — APPROVED** [review]
LGTM 🚀 

Ideally, fixing the test skip could clarify the motivation why we had to do so. At least, doing so in a follow up would be nice :)

**Pull request merged**
